src/analog.rs

```diff
@@ -0,0 +1,50 @@
+//! Analog-digital conversion traits
+
+use nb;
+
+/// This trait denotes an object, e.g. a GPIO pin, that is configured for use as an ADC channel.
+/// Objects of this type are passed to ADC-mode-specific methods.
+pub trait AdcChannel {
+    /// Channel ID type
+    ///
+    /// e.g. u8 for ADC channels 1-n, or a (partid, pinid) tuple for something more complex.
+    type ID;
+
+    /// ID of the channel associated with this pin
+    const CHANNEL: Self::ID;
+}
+
+/// ADCs that sample on single channels per request, and do so at the time of the request.
+pub trait SingleMode<Word, Pin: AdcChannel> {
+    /// Error type returned by ADC methods
+    type Error;
+
+    /// Take an analog reading from the specified pin
+    fn read_channel(&mut self, pin: &mut Pin) -> nb::Result<Word, Self::Error>;
+}
+
+/// ADCs that sample continuously without requiring requests
+pub trait Continuous<Word, Pin: AdcChannel> {
+    /// Error type returned by ADC methods
+    type Error;
+
+    /// Take ownership of an ADC channel/pin
+    fn add_channel(&mut self, pin: Pin) -> nb::Result<(), Self::Error>;
+
+    /// Release the ADC channel/pin
+    fn release_channel(&mut self) -> nb::Result<Pin, Self::Error>;
+
+    /// Start a continuous conversion cycle
+    fn start(&mut self) -> nb::Result<(), Self::Error>;
+
+    /// Stop conversions.
+    ///
+    /// If the conversions cannot be stopped immediately, returns nb::Error::WouldBlock.
+    fn stop(&mut self) -> nb::Result<(), Self::Error>;
+
+    /// Read the specified channel
+    fn read(&mut self, chan: Pin::ID) -> nb::Result<Word, Self::Error>;
```

Comment:
I'm confused as to how this API ought to work -- maybe I'm missing something.

It seems as though the intended flow would be something like:

```Rust
fn get_samples() {
    let pin = Channel::new(); // actually get a channel though
    let adc = Adc::init(); // actually get an ADC handle

    adc.add_channel(pin);
    adc.start();
    for _ in 0..=10 {
        let _ = adc.read(pin); // fails borrow check
    }
    adc.stop();
    let rpin = adc.release_channel().unwrap();
}
```

I'm omitting error checking, etc, and might have made a mistake with this hasty example. The fact remains, however, that unless I'm misinterpreting when and how `add_channel()` and `read()` must interact, this interface cannot work.

Reply:
I'm also confused as to how this would work with multiple continuously-sampled channels.

Reply:
Whoops, you're totally right. `read` should, I think, probably take an `AdcChannel::CHANNEL`.

Reply:
Sorry for the earlier deleted comments - I misunderstood your question.

Note that `read` takes a `Pin::ID` argument, not the `Pin` itself. So the code would look something like:
```rust
let pin: gpio::PA0<Analog> = gpioa.PA0.into_analog(); // PA0<Analog> impls AdcChannel<ADC1>
let adc = adc::init(...);
adc.add_channel(pin);
adc.start();
for _ in 0..=10 {
    let _ = adc.read(PA0::CHANNEL);
}
adc.stop();
let rpin = adc.release_channel(PA0::CHANNEL).unwrap();
```